src/eduid/webapp/idp/other_device.py

```diff
@@ -0,0 +1,229 @@
+        request_id: Optional[str],
+        ip_address: str,
+        user_agent: Optional[str],
+        ttl: timedelta,
+        reauthn_required: bool = False,
+    ) -> OtherDevice:
+        _uuid = uuid.uuid4()
+        short_code = make_short_code()
+        now = utc_now()
+        return cls(
+            state_id=OtherDeviceId(str(_uuid)),
+            state=OtherDeviceState.NEW,
+            short_code=short_code,
+            eppn=eppn,
+            device1=Device1Data(
+                ref=device1_ref,
+                authn_context=authn_context,
+                request_id=request_id,
+                ip_address=ip_address,
+                user_agent=user_agent,
+                reauthn_required=reauthn_required,
+            ),
+            device2=Device2Data(),
+            created_at=now,
+            expires_at=now + ttl,
+        )
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return self.dict()
+
+    def to_json(self):
+        """ For debug logging ONLY. Redacts the response code if set. """
+        data = self.to_dict()
+        if data['device2']['response_code']:
+            data['device2']['response_code'] = 'REDACTED'
+        return json.dumps(data, indent=4, cls=EduidJSONEncoder)
+
+    @classmethod
+    def from_dict(cls: Type[OtherDevice], data: Mapping[str, Any]) -> OtherDevice:
+        return cls(**data)
+
+
+def make_short_code() -> str:
+    digits = int.from_bytes(os.urandom(4), byteorder='big') % 1000000
+    return '{:06d}'.format(digits)
+
+
+class OtherDeviceDB(BaseDB):
+    def __init__(self, db_uri: str, db_name: str = 'eduid_idp', collection: str = 'other_device'):
+        super().__init__(db_uri, db_name, collection=collection)
+
+        indexes = {
+            'auto-discard': {'key': [('expires_at', 1)], 'expireAfterSeconds': 0},
+            'unique-state-id': {'key': [('state_id', 1)], 'unique': True},
+        }
+        self.setup_indexes(indexes)
+
+    def save(self, state: OtherDevice) -> bool:
+        """
+        Add a new OtherDevice to the database, or update an existing one.
+        """
+        result = self._coll.replace_one({'_id': state.obj_id}, state.to_dict(), upsert=True)
+        logger.debug(
+            f'Saved OtherDevice {state} in the db: '
+            f'matched={result.matched_count}, modified={result.modified_count}, upserted_id={result.upserted_id}'
+        )
+        return result.acknowledged
+
+    def get_state_by_id(self, state_id: OtherDeviceId) -> Optional[OtherDevice]:
+        state = self._get_document_by_attr('state_id', str(state_id))
+        if not state:
+            logger.debug(f'Other-device state with state_id {state_id} not found in the database')
+            return None
+        return OtherDevice.from_dict(state)
+
+    def add_new_state(self, ticket: 'LoginContext', user: Optional[User], ttl: timedelta) -> OtherDevice:
+        from eduid.webapp.idp.assurance import get_requested_authn_context
```

Comment:
I implemented this in commit cc2f06dc3e7152d0c6d0031d5aedfae18bae905e, have a look. As I said in the commit message, I'm not sure I think it is an improvement because it spreads handling of assurances to different parts of the code.